checker/src/org/checkerframework/checker/nullness/KeyForValue.java

```diff
@@ -0,0 +1,113 @@
+ * is used in {@link KeyForAnnotatedTypeFactory#isKeyForMap(String, ExpressionTree)}.
+ */
+public class KeyForValue extends CFAbstractValue<KeyForValue> {
+    /**
+     * If the underlying type is a type variable or a wildcard, then this is list of maps this for
+     * which this value is a key. Otherwise, it's null.
+     */
+    Set<String> keyForMaps = null;
+
+    public KeyForValue(
+            CFAbstractAnalysis<KeyForValue, ?, ?> analysis,
+            Set<AnnotationMirror> annotations,
+            TypeMirror underlyingType) {
+        super(analysis, annotations, underlyingType);
+        AnnotationMirror keyfor = AnnotationUtils.getAnnotationByClass(annotations, KeyFor.class);
+        if (keyfor != null
+                && (underlyingType.getKind() == TypeKind.TYPEVAR
+                        || underlyingType.getKind() == TypeKind.WILDCARD)) {
+            keyForMaps = new LinkedHashSet<>();
+            List<String> list =
+                    AnnotationUtils.getElementValueArray(keyfor, "value", String.class, true);
+            keyForMaps.addAll(list);
+        }
+    }
+
+    @Override
+    public KeyForValue leastUpperBound(KeyForValue other) {
+        KeyForValue lub = super.leastUpperBound(other);
+        if (other == null || other.keyForMaps == null || this.keyForMaps == null) {
+            return lub;
+        }
+        // Lub the keyFoMaps by intersecting the sets.
+        lub.keyForMaps = new LinkedHashSet<>();
+        lub.keyForMaps.addAll(this.keyForMaps);
+        lub.keyForMaps.retainAll(other.keyForMaps);
+        if (lub.keyForMaps.isEmpty()) {
+            lub.keyForMaps = null;
+        }
+        return lub;
+    }
+
+    @Override
+    public KeyForValue mostSpecific(KeyForValue other, KeyForValue backup) {
+        KeyForValue mostSpecific = super.mostSpecific(other, backup);
+        if (mostSpecific == null) {
+            // mostSpecific is null if the two types are not comparable.  This is normally
+            // because one of this or other is a type variable and annotations is empty, but the
+            // other annotations are not empty.  In this case copy the keyForMaps and
+            if (other == null) {
+                return this;
+            }
+            if (other.getAnnotations().isEmpty()) {
+                other.addKeyFor(this.keyForMaps);
+                return other;
+            } else if (this.getAnnotations().isEmpty()) {
+                this.addKeyFor(other.keyForMaps);
+                return this;
+            }
+            return null;
+        }
+
+        mostSpecific.addKeyFor(this.keyForMaps);
+        if (other != null) {
+            mostSpecific.addKeyFor(other.keyForMaps);
+        }
+        return mostSpecific;
+    }
+
+    private void addKeyFor(Set<String> newKeyForMaps) {
+        if (newKeyForMaps == null) {
+            return;
+        }
+        if (keyForMaps == null) {
+            keyForMaps = new LinkedHashSet<>();
```

Comment:
Slightly restructure the logic to not allocate the linked list and then set the field back to null.